### ISSUE TITLE
keywordize keys fn

### ISSUE THREAD

**Issue by jacekschae** [ovotech/ring-jwt#11]
Hi there, thanks for this great library. I have one thing that would be interested to know if you would be up for discussion. 

In the `token.clj` file on line [19](https://github.com/ovotech/ring-jwt/blob/9b0aaebf036a5335af8d0fe49e3e56f6e53e9926/src/ring/middleware/token.clj#L19) a function `keywordize-keys` is used. 

Taking into account that that the names we put in `claims` should be namespaced i.e internet address - http://example.com/user_authorization the outcome is non-valid keyword - 
```clojure
:http://example.com/user_authorization
```

would it be possible to remove this function and operate on strings or provide a flag to keywordize keys?

Thanks

**Comment by kelveden** [comment]
@jacekschae that's a fair point. There's [nothing specific](https://tools.ietf.org/html/rfc7519#section-4.3) in the JWT specification that explicitly addresses how to avoid collisions in private claims but you're right that just blindly keywordizing isn't going to work for everyone.

There are two options as far as I can see:

1. Add an option to the configuration that explicitly "switches off" keywordizing of the claims.
2. Don't keywordize keys that contain `/` - the implication being that claim names with a "/" must be using that "/" to namespace.

My preference is for (2) I think. Will look into it.

**Comment by jacekschae** [comment]
@kelveden thanks for looking into this!

It's not explicitly stated there but one can assume by reading the 4.2
```
4.2.  Public Claim Names

   Claim Names can be defined at will by those using JWTs.  However, in
   order to prevent collisions, any new Claim Name should either be
   registered in the IANA "JSON Web Token Claims" registry established
   by Section 10.1 or be a Public Name: a value that contains a
   Collision-Resistant Name.  In each case, the definer of the name or
   value needs to take reasonable precautions to make sure they are in
   control of the part of the namespace they use to define the Claim
   Name.
```

> From Auth0 - [Create Namespaced Custom Claims](https://auth0.com/docs/tokens/guides/create-namespaced-custom-claims):
> Namespaces are arbitrary identifiers, so technically you can call your namespace anything you want. However, using the URI of a resource you control is conventional (following the way XML namespaces are defined).
> 
> Example namespace: `http://www.myexample.com/`
> 
> Because a URI must be unique and because this URI is under your control, you can generally avoid the risk that someone else is using the same namespace. You can also optionally store the definition of your namespace at the URI, making your namespace self-documenting.

3. Wouldn't be better to not keywordize when a string contains `http://` or `https://` ? `keywords` can be `namespaced` - i.e. `:namespaced/keyword`

**Comment by kelveden** [comment]
Regarding keywords, yeah I did wonder about how far to take the logic in specifically matching URIs but concluded in the end that just detecting the "/" was enough - most folks are just going to use URIs as the logical way to namespace claims anyway. It's only going to be Clojure folks who might consider namespacing something like `a/b`.

The logic would potentially get deceptively complicated too - you've got to consider how to match a URI (presumably with a regex). And then converting a string of the form "a/b" to a namespaced keyword isn't as simple as just calling `(keyword "a/b")` - one has to split out the namespace part into `(keyword "a" "b")`.

So, for now, I've kept it simple. It can always be changed later if there's enough interest. Worst comes to the worst, folks can just pass the claims through `clojure.walk/keywordize-keys` themselves.
